java-checks/src/main/java/org/sonar/java/checks/regex/ImpossibleBoundariesCheck.java

```diff
@@ -20,19 +20,31 @@
 package org.sonar.java.checks.regex;
 
 import java.util.Collections;
+import java.util.HashSet;
+import java.util.Set;
+
 import org.sonar.check.Rule;
+import org.sonar.java.checks.helpers.RegexReachabilityChecker;
 import org.sonar.java.checks.helpers.RegexTreeHelper;
 import org.sonar.java.regex.RegexParseResult;
 import org.sonar.java.regex.ast.AutomatonState;
 import org.sonar.java.regex.ast.BoundaryTree;
+import org.sonar.java.regex.ast.DisjunctionTree;
 import org.sonar.java.regex.ast.LookAroundTree;
 import org.sonar.java.regex.ast.RegexBaseVisitor;
+import org.sonar.java.regex.ast.RegexTree;
 import org.sonar.plugins.java.api.tree.ExpressionTree;
 
+import static org.sonar.java.regex.ast.AutomatonState.TransitionType.CHARACTER;
+
 @Rule(key = "S5996")
 public class ImpossibleBoundariesCheck extends AbstractRegexCheck {
 
   private static final String MESSAGE = "Remove or replace this boundary that will never match because it appears %s mandatory input.";
+  private static final String SOFT_MESSAGE =
+    "Remove or replace this boundary that can only match if the previous part matched the empty string because it appears %s mandatory input.";
+  private final Set<RegexTree> excluded = new HashSet<>();
+  private final RegexReachabilityChecker regexReachabilityChecker = new RegexReachabilityChecker(false);
```

Comment:
This checker should have its cache cleared when visiting a new regex (i.e. on every call to `checkRegex`). Having the states of the old regex in the cache won't buy us anything (it will never lead to a cache hit) and just increase the chance that we hit the cache size limit.